ubuntu/main.cc

```diff
@@ -0,0 +1,257 @@
+  event.phase = phase;
+  event.x = x;
+  event.y = y;
+  event.timestamp =
+      std::chrono::duration_cast<std::chrono::microseconds>(
+          std::chrono::high_resolution_clock::now().time_since_epoch())
+          .count();
+  FlutterEngineSendPointerEvent(
+      reinterpret_cast<FlutterEngine>(glfwGetWindowUserPointer(window)), &event,
+      1);
+}
+
+static void GLFWcursorPositionCallback(GLFWwindow *window, double x, double y) {
+  GLFWcursorPositionCallbackAtPhase(window, FlutterPointerPhase::kMove, x, y);
+}
+
+static void GLFWmouseButtonCallback(GLFWwindow *window, int key, int action,
+                                    int mods) {
+  double x, y;
+  if (key == GLFW_MOUSE_BUTTON_1 && action == GLFW_PRESS) {
+    glfwGetCursorPos(window, &x, &y);
+    GLFWcursorPositionCallbackAtPhase(window, FlutterPointerPhase::kDown, x, y);
+    glfwSetCursorPosCallback(window, GLFWcursorPositionCallback);
+  }
+  if (key == GLFW_MOUSE_BUTTON_1 && action == GLFW_RELEASE) {
+    glfwGetCursorPos(window, &x, &y);
+    GLFWcursorPositionCallbackAtPhase(window, FlutterPointerPhase::kUp, x, y);
+    glfwSetCursorPosCallback(window, nullptr);
+  }
+}
+
+static void GLFWKeyCallback(GLFWwindow *window, int key, int scancode,
+                            int action, int mods) {
+  if (key == GLFW_KEY_ESCAPE && action == GLFW_PRESS) {
+    glfwSetWindowShouldClose(window, GLFW_TRUE);
+  }
+}
+
+static void GLFWwindowSizeCallback(GLFWwindow *window, int width, int height) {
+  FlutterWindowMetricsEvent event = {};
+  event.struct_size = sizeof(event);
+  event.width = width;
+  event.height = height;
+  event.pixel_ratio = 1.0;
+  FlutterEngineSendWindowMetricsEvent(
+      reinterpret_cast<FlutterEngine>(glfwGetWindowUserPointer(window)),
+      &event);
+}
+
+static bool GLFWMakeContextCurrent(void *user_data) {
+  GLFWwindow *window = reinterpret_cast<GLFWwindow *>(user_data);
+  glfwMakeContextCurrent(window);
+  return true;
+}
+
+static bool GLFWClearContext(void *user_data) {
+  glfwMakeContextCurrent(nullptr);
+  return true;
+}
+
+static bool GLFWPresent(void *user_data) {
+  GLFWwindow *window = reinterpret_cast<GLFWwindow *>(user_data);
+  glfwSwapBuffers(window);
+  return true;
+}
+
+static uint32_t GLFWGetActiveFbo(void *user_data) { return 0; }
+
+// Clears the GLFW window to Material Blue-Grey.
+//
+// This function is primarily to fix an issue when the Flutter Engine is
+// spinning up, wherein artifacts of existing windows are rendered onto the
+// canvas for a few moments.
+//
+// This function isn't necessary, but makes starting the window much easier on
+// the eyes.
+static void GLFWClearCanvas(GLFWwindow *window) {
+  glfwMakeContextCurrent(window);
+  // This color is Material Blue Grey.
+  glClearColor(236, 239, 241, 0);
+  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+  glFlush();
+  glfwSwapBuffers(window);
+  glfwMakeContextCurrent(nullptr);
+}
+
+static bool RunFlutterEngine(GLFWwindow *window,
```

Comment:
Functions other than the self-explanatory GLFW callback functions should have declaration comments; code in this repo should follow the Google style guide.

Reply:
Fixed!